Controllers management

vicalro opened the pull request:
This PR adds the features of adding, removing and listing controllers to LSIs trough the REST plugin.
Examples on how to use it can be found in the xcli tool.

In order to use the list feature, the API of rofl-common needs to be extended. You may find an extension in vicalro/rofl-common@3d251d6fe5dc4acb3012e114ba2fe823d17aaf67 (branch `list_controllers`)

vicalro committed:
Add controller from rest plugin implemented

vicalro committed:
Basic remove controller functionality for REST interface

vicalro committed:
implementation of list in xcli and in endpoint. It is still not glued in the REST plugin

vicalro committed:
Implemented list IDs from LSI

vicalro committed:
Implemented ID assigned when adding a new controller

vicalro committed:
Added warnings on add and remove controller malformed commands

vicalro committed:
Implementation of REST call show controller

vicalro committed:
Added examples exec add/remove controller to xcli

vicalro committed:
Changed the structure of the information of the controllers for better parsing

vicalro committed:
Erased "rpc" from calls to the switch manager. Using only read lock when gathering info of controllers

vicalro committed:
Small fix in label of controller information

vicalro committed:
Added Equal as a role for controllers and using get_role() in crofbase to get the role

vicalro committed:
Cleaning old commnents

vicalro committed:
Merge branch 'integration-0.8' into controllers_management

vicalro committed:
Following previuos order in register of REST calls

vicalro committed:
Changed the way we handle errors in management of controllers with REST

vicalro committed:
Removed unnecessary spaces

vicalro committed:
Deleted unnecessary stringstreams

msune commented:
List controllers and show controller info should not need `-m` option:

```
xcli(local)>show lsi dp0 controllers
ERROR(401): HTTP/GET 'http://127.0.0.1:5757/info/lsi/dp0/controllers'
Unauthorised; is mgmt support enabled (-m)?

xcli(local)>show lsi dp0 controller 0
ERROR(401): HTTP/GET 'http://127.0.0.1:5757/info/lsi/dp0/controller/0'
Unauthorised; is mgmt support enabled (-m)?

xcli(local)>
```

msune commented:
```
xcli(local)>show lsi dp0 controller 0
{
    "Controller" : {
        "id" : 0,
        "channel-status" : "not-established",
        "mode" : "equal",
        "connections" : {
            "0" : {
                "protocol-type" : "plain",
                "ip" : "127.0.0.1",
                "port" : 6653
            }
        }
    }
}

xcli(local)>
```

`"Controller"` -> `"controller"`

And:

```
xcli(local)>exec add controller dp0 tcp:192.168.1.1:6633
{                                                        
    "assigned id" : 1                                    
}

```

I think for consistency (list controllers) should be:

```
xcli(local)>exec add controller dp0 tcp:192.168.1.1:6633
{                                                        
    "controller-id" : 1
}

```

msune commented:
Exec with bad parameters dos not send an HTTP 500 error, although it fails to add it:

```
xcli(local)>exec add controller dp0 tcp:192.168.1.1::6633

xcli(local)>show lsi dp0 controllers
{
    "controller-ids" : [
    ]
}

xcli(local)>exec add controller dp0 tcp192.168.1.1::6633

xcli(local)>show lsi dp0 controllers 
{
    "controller-ids" : [
    ]
}

xcli(local)>exec add controller dp0 tcp45654654192.168.1.1::6633

xcli(local)>show lsi dp0 controllers
{
    "controller-ids" : [
    ]
}
```

msune commented:
New APIs are not listed in the index (GET to root):

e.g. http://10.0.0.2:5757/

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -236,6 +236,102 @@ void lsi_groups(const http::server::request &req,
+	json_spirit::Object table;
+
+	//Perform security checks
+        if(!authorised(req,rep)) return;
```
extra space. Authorisization here is not needed

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -236,6 +236,102 @@ void lsi_groups(const http::server::request &req,
+	json_spirit::Object j_wrap, j_ctl, j_list;
+
+	//Perform security checks
+        if(!authorised(req,rep)) return;
```
extra space. Authorisation here not needed.

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -236,6 +236,102 @@ void lsi_groups(const http::server::request &req,
+	try{
+		switch_manager::get_controller_info(dpid, ctl_id, ctl_info);
+	}catch(...){
+		throw "Unable to get Controller info for lsi";
```
Shouldn't this go as part of the `rep.content`?

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -358,6 +454,87 @@ void create_lsi(const http::server::request &req, http::server::reply &rep, boos
+	uint64_t assigned_id;
+
+	//Perform security checks
+        if(!authorised(req,rep)) return;
```
extra space

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -358,6 +454,87 @@ void create_lsi(const http::server::request &req, http::server::reply &rep, boos
+	std::string lsi_name = std::string(grps[1]);
+
+	//Check if LSI exists;
+        if(!switch_manager::exists_by_name(lsi_name)){
```
space

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -358,6 +454,87 @@ void create_lsi(const http::server::request &req, http::server::reply &rep, boos
+	try{
+		assigned_id = switch_manager::connect_to_ctl(dpid, socket_type, socket_params);
+	}catch(...){
+                throw "Unable to add controller to lsi";
```
wrong identation. This should go to `rep.content`

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -395,6 +572,42 @@ void destroy_switch(const http::server::request &req, http::server::reply &rep,
+	ctlid_str = std::string(grps[2]);
+
+	//Check if LSI exists;
+        if(!switch_manager::exists_by_name(lsi_name)){
```
space

msune commented on src/xdpd/management/plugins/rest/sw-controllers.cc:
```diff
@@ -395,6 +572,42 @@ void destroy_switch(const http::server::request &req, http::server::reply &rep,
+	try{
+		switch_manager::disconnect_from_ctl(dpid, ctlid);
+	}catch(...){
+                throw "Unable to remove controller from lsi";
```
idem as before. indentation as should be the return  body of HTTP

msune commented on src/xdpd/management/plugins/rest/xcli:
```diff
@@ -302,6 +339,29 @@ def execute(socket, command, req_confirm=False):
+			else:
+				return "Usage: exec add controller <lsi_name> <mode:ip:port>"
+
+		elif "remove" in command:
```
This entire block, until line 364 is indented with spaces. Please move it to hard tabs.

(For the record: we don't care about PEP8, tabs were created to indent!)

vicalro committed:
Polish of REST management of controllers

- List and show controllers don't require authorized access
- Fixed return message after adding a new controller
- Capturing exception when parsing the parameters of the new controller in xcli
- New APIs listed in the index

vicalro commented:
Thanks for the comments. Issues fixed. Check it out!

vicalro committed:
Fixed indentation and removed unnecessary try/catches

msune commented:
ACK. But we need to merge the list_controllers into rofl-common, but first https://github.com/bisdn/rofl-common/issues/17 needs to be fixed, since the commit needs to go on top of integration-0.6 as agreed.

Holding it.
